File: J9Ch/src/S_5_Dynamic_Problem_I/Optional_7/python_cpp/_396_Coins_in_a_Line_III_Hard.cpp

class Solution {
public:
    /**
     * @param values: a vector of integers
     * @return: a boolean which equals to true if the first player will win
     */
    bool firstWillWin(vector<int> &values) {
        // write your code here
        int n = values.size();
        vector<int> sum(values);
        vector<vector<int> > f(n, vector<int>(n));
        for (int i=1; i<n; ++i) sum[i] += sum[i-1];
        for (int j=0; j<n; ++j) f[0][j] = values[j];
        for (int i=1; i<n; ++i) {
            f[i][0] = max(values[0]+sum[i]-sum[0]-f[i-1][1], values[i]+sum[i-1]-f[i-1][0]);
            for (int j=1; j+i<n; ++j)
                f[i][j] = max(values[j]+sum[j+i]-sum[j]-f[i-1][j+1], values[j+i]+sum[j+i-1]-sum[j-1]-f[i-1][j]);
        }
        if (f[n-1][0]<sum[n-1]-f[n-1][0]) return false;
        else return true;
    }
};

File: J9Ch/src/S_6_Dynamic_Problem_II/Required_10/python_cpp/_396_Coins_in_a_Line_III_Hard.py

class Solution:
    # @param values: a list of integers
    # @return: a boolean which equals to True if the first player will win
    def firstWillWin(self, values):
        # write your code here
        n = len(values)
        sumv, f = [], []
        for i in xrange(n): sumv.append(values[i])
        for i in xrange(1, n): sumv[i] += sumv[i-1]
        f.append([])
        for j in xrange(n): f[0].append(values[j])
        for i in xrange(1, n):
            f.append([])
            f[i].append(max(values[0]+sumv[i]-sumv[0]-f[i-1][1], values[i]+sumv[i-1]-f[i-1][0]))
            for j in xrange(1, n-i):
                f[i].append(max(values[j]+sumv[j+i]-sumv[j]-f[i-1][j+1], values[j+i]+sumv[j+i-1]-sumv[j-1]-f[i-1][j]))
        if f[n-1][0]<sumv[n-1]-f[n-1][0]: return False
        else: return True